Add iterable checking for response content

qiuzhong opened the pull request:
We need check if the response content is iterable before we iterable
over it. Otherwise, it might cause the server hanging. See #8368

qiuzhong committed:
Add iterable checking for response content

We need check if the response content is iterable before we iterable
over it. Otherwise, it might cause the server hanging. See #8368

Ms2ger commented on tools/wptserve/wptserve/response.py:
```diff
@@ -183,6 +183,9 @@ def iter_content(self, read_file=False):
         True, the entire content of the file will be returned as a string facilitating
         non-streaming operations like template substitution.
         """
+        if not isinstance(self.content, Iterable) and not hasattr(self.content, "read"):
```
I don't think the `Iterable` check works in general.

gsnedders replied:
No, it definitely doesn't. The vast majority of iterables don't descend from it.

qiuzhong replied:
[Python documentation](https://docs.python.org/2.7/library/collections.html#collections.Iterable) says 
> `collections.Iterable`: ABC for classes that provide the `__iter__()` method. 

`isinstance(xxx, collections.Iterable)` returns `True` for all the sequence types and `dict` types and `file` objects. 
It's not the best way for checking iterable. Maybe I need to remove this `if` statement and add a `try...catch` statement whenever calling the `item_content` method.

Ms2ger commented on tools/wptserve/wptserve/response.py:
```diff
@@ -183,6 +183,9 @@ def iter_content(self, read_file=False):
         non-streaming operations like template substitution.
         """
+        if not isinstance(self.content, Iterable) and not hasattr(self.content, "read"):
+            raise TypeError("Response content must be an iterable or file-like object!")
```
Raising an exception here doesn't work; that only works inside the `handler()` call.

```
________________ TestFunctionHandler.test_non_iterable_content _________________
self = <functional.test_handlers.TestFunctionHandler testMethod=test_non_iterable_content>
    def test_non_iterable_content(self):
        @wptserve.handlers.handler
        def handler(request, response):
            response.content = 1234
            return response
    
        route = ("GET", "/non_iterable_rv", handler)
        self.server.router.register(*route)
        resp = self.request(route[1])
>       self.assertEqual(500, resp.getcode())
E       AssertionError: 500 != 200
wptserve/tests/functional/test_handlers.py:205: AssertionError
----------------------------- Captured stderr call -----------------------------
DEBUG:web-platform-tests:Route pattern: ^/(.*)$
DEBUG:web-platform-tests:Route pattern: ^/(.*\.asis)$
DEBUG:web-platform-tests:Route pattern: ^/(.*\.py)$
DEBUG:web-platform-tests:Using default configuration
INFO:web-platform-tests:Starting http server on localhost:35117
DEBUG:web-platform-tests:Route pattern: ^/non\_iterable\_rv$
DEBUG:web-platform-tests:GET /non_iterable_rv
DEBUG:web-platform-tests:Found handler FunctionHandler
DEBUG:web-platform-tests:200 GET /non_iterable_rv (None) 0
ERROR:web-platform-tests:Traceback (most recent call last):
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/server.py", line 544, in handle_one_request
    self.finish_handling_h1(request_line_is_valid)
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/server.py", line 243, in finish_handling_h1
    self.finish_handling(request, response, handler)
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/server.py", line 278, in finish_handling
    response.write()
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/response.py", line 223, in write
    self.write_content()
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/response.py", line 217, in write_content
    for item in self.iter_content():
  File "/home/travis/build/web-platform-tests/wpt/tools/wptserve/wptserve/response.py", line 187, in iter_content
    raise TypeError("Response content must be an iterable or file-like object!")
TypeError: Response content must be an iterable or file-like object!
```

qiuzhong replied:
What I thought was any non-interable response content will raise a `TypeError` and an HTTP request gets a `500` status error. I assigned a number to `response.content` in handler on purpose. But obviously, it didn't work. Any good solution suggestion?

Sorry for this PR that didn't pass the unit test. I failed to set up CI environment before. My local CI environment for tools unit tests is set up successfully now. I'll run more unit test before I file any PRs.

Ms2ger commented on tools/wptserve/wptserve/response.py:
```diff
@@ -183,6 +183,9 @@ def iter_content(self, read_file=False):
         True, the entire content of the file will be returned as a string facilitating
         non-streaming operations like template substitution.
         """
+        if not isinstance(self.content, Iterable) and not hasattr(self.content, "read"):
```
Also, this check should go in the `else` block below, IMO.

Ms2ger commented on tools/wptserve/wptserve/response.py:
```diff
@@ -183,6 +183,9 @@ def iter_content(self, read_file=False):
         non-streaming operations like template substitution.
         """
+        if not isinstance(self.content, Iterable) and not hasattr(self.content, "read"):
+            raise TypeError("Response content must be an iterable or file-like object!")
```
Also, this yields a pretty useless stack trace. Can we turn the response attribute into a getter/setter and check in the setter?

qiuzhong committed:
Check response content before write it

The response content requires an iterable. When a bad response handler
assigns incorrect type data to the response, calling `response.write()`
will raise a `TypeError` the testing server crashes. We need to check
if the response content is iterable before the response writes its content.

qiuzhong commented:
This PRs derives from #8638. In its comments, @jgraham suggested that  
> we should have a check that the content actually is iterable before we try to iterate over it and in case it is not raise a `ValueError`.

Indeed, an inappropriate handler (on purpose) like this
```
def handler(request, response):
	response.content = 1
	return response.content
```
sets the response content an int type, which is not be used for iteration. Such case was not handled well by `tools/wptserve/wptserve/server.py::BaseWebTestRequestHandler.finish_handling`: calling `response.write()` will finally raise a `TypeError` and the testing server crashes. This PR tries to check if the response content iterable before the response writes its content. 
I added the iterable checking in `iter_content()` method before, but it didn't work. So I removed it completely  and used a new way to check if the response content is an iterable.

qiuzhong commented:
@Ms2ger, @gsnedders, @jgraham , would you please take a look at it and maybe give some suggestion?
